src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
+    static let waitHandleCollection = ConcurrentDictionary<_,_>() 
+    static let unregisterTimer = new System.Threading.Timer( TimerCallback( ThreadPoolWait.ToUnregister ), null, 10, 10 )
+    static member WaitForHandleWithCancellation (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) (token:CancellationToken) =
+        token.Register( fun _ -> ThreadPoolWait.TryWakeup handle |> ignore ) |> ignore 
+        ThreadPoolWait.WaitForHandle infoFunc handle continuation unblockHandle
+    static member WaitForHandle (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) =
+        if handle.WaitOne(0) then 
+            continuation() 
+            if Utils.IsNotNull unblockHandle then 
+                // If there is an unblock handle, set it. 
+                unblockHandle.Set() |> ignore
+        else
+            /// Uniquely identify this async job and its resource removal. 
+            let jobObject = Object()
+            let rwh = ThreadPool.RegisterWaitForSingleObject( handle, new WaitOrTimerCallback(ThreadPoolWait.CallBack), (infoFunc,handle,continuation,unblockHandle,jobObject) , -1, true )
+            waitHandleCollection.Item(jobObject) <- ( rwh, infoFunc, ref DateTime.MinValue.Ticks, handle )
+    static member private CallBack( state: Object ) (timeout:bool) = 
+        if not timeout then
+            try
+                let infoFunc,handle,continuation,unblockHandle,jobObject = state :?> ((unit->string)*WaitHandle*(unit->unit)*ManualResetEvent*Object)
+                continuation() 
+                if Utils.IsNotNull unblockHandle  then 
+                    // If there is an unblock handle, set it. 
+                    unblockHandle.Set() |> ignore  
+                let bExist, tuple = waitHandleCollection.TryGetValue( jobObject )
+                if bExist then 
+                    let rwh, _, _, _ = tuple 
+                    rwh.Unregister(null) |> ignore 
+                    waitHandleCollection.TryRemove( jobObject ) |> ignore 
+                else
+                    rwhToRemove.Enqueue( jobObject )
+            with 
+            | ex -> 
+                Logger.LogF( LogLevel.Error, fun _ -> sprintf "ThreadPoolWait.Callback, exception: %A" ex )
+                reraise()
+        else
+            failwith "ThreadPoolWait.Callback, timeout path reached (impossible execution route)"
+    static member ToUnregister(o:Object) = 
+        let requeue = List<_>()        
+        while not rwhToRemove.IsEmpty do
+            let bRemove, jobObject = rwhToRemove.TryDequeue()
+            if bRemove then 
+                let bExist, tuple = waitHandleCollection.TryGetValue( jobObject )
+                if bExist then 
+                    let rwh, _, _, _ = tuple
+                    rwh.Unregister(null) |> ignore 
+                else
+                    requeue.Add( jobObject )
```

Comment:
in what scenario that it is possible that the jobObjects exists in rwhToRemove but not in waitHandleCollection?  If happens, should this be an error that we should log and investigate?